File: cpp/src/jvm/method_walker.cpp
#include "method_walker.hpp"
#include "walker.hpp"
#include <cstring>
#include <stdexcept>

namespace marrow {

// Minimal CP resolver: reads the raw slot as Symbol* and decodes to UTF-8.
// Lives here in addition to the one in field_reader.cpp to keep the two
// modules independent.
namespace {
class CPNames {
public:
    CPNames(VMMeta* vm, uint64_t cp_ptr)
        : vm_(vm), r_(vm->reader()), cp_(cp_ptr)
    {
        if (auto* t = vm->type("ConstantPool")) {
            header_size_ = t->size;
            if (t->has_field("_length"))
                length_ = r_->read_i32(cp_ptr + t->field("_length")->offset);
        }
    }
    std::string symbol(int idx) const {
        if (!cp_ || idx <= 0 || idx >= length_) return {};
        try {
            uint64_t sym = r_->read_u64(cp_ + header_size_ + size_t(idx) * 8);
            if (!sym || sym < 0x10000) return {};
            return read_symbol(vm_, sym);
        } catch (...) { return {}; }
    }
private:
    VMMeta* vm_;
    Reader* r_;
    uint64_t cp_;
    size_t header_size_ = 0;
    int32_t length_ = 0;
};
}

static std::vector<uint64_t> array_methods(VMMeta* vm, uint64_t klass_ptr)
{
    const TypeInfo* ik = vm->type("InstanceKlass");
    uint64_t arr = vm->reader()->read_u64(klass_ptr + ik->field("_methods")->offset);
    if (!arr) return {};
    int32_t length = vm->reader()->read_i32(arr);
    if (length <= 0) return {};
    auto raw = vm->reader()->read(arr + 8, size_t(length) * 8);
    std::vector<uint64_t> out(length);
    std::memcpy(out.data(), raw.data(), raw.size());
    return out;
}

std::vector<MethodSnapshot> methods_of(VMMeta* vm, uint64_t klass_ptr)
{
    Reader* r = vm->reader();
    const TypeInfo* mt = vm->type("Method");
    const TypeInfo* cmt = vm->type("ConstMethod");
    size_t off_constmethod = mt->field("_constMethod")->offset;
    size_t off_name_idx    = cmt->field("_name_index")->offset;
    size_t off_sig_idx     = cmt->field("_signature_index")->offset;
    size_t off_code_size   = cmt->field("_code_size")->offset;
    size_t off_cp          = cmt->field("_constants")->offset;
    size_t header_size     = cmt->size;

    std::vector<MethodSnapshot> out;
    for (uint64_t m : array_methods(vm, klass_ptr)) {
        if (!m) continue;
        uint64_t cm = r->read_u64(m + off_constmethod);
        if (!cm) continue;
        uint64_t cp_ptr = r->read_u64(cm + off_cp);
        CPNames cp(vm, cp_ptr);
        std::string name = cp.symbol(r->read_u16(cm + off_name_idx));
        std::string sig  = cp.symbol(r->read_u16(cm + off_sig_idx));
        uint16_t code_size = r->read_u16(cm + off_code_size);
        out.push_back({m, cm, std::move(name), std::move(sig),
                       code_size, cm + header_size});
    }
    return out;
}

std::optional<MethodSnapshot>
find_method(VMMeta* vm, uint64_t klass_ptr,
             const std::string& name,
             const std::optional<std::string>& signature)
{
    for (auto& m : methods_of(vm, klass_ptr)) {
        if (m.name != name) continue;
        if (signature && m.signature != *signature) continue;
        return m;
    }
    return std::nullopt;
}

void rewrite_method_body(VMMeta* vm, const MethodSnapshot& method,
                          const std::vector<uint8_t>& new_bytecode)
{
    if (new_bytecode.size() > method.code_size)
        throw std::runtime_error("new bytecode exceeds code_size slot");
    Reader* r = vm->reader();
    const TypeInfo* mt = vm->type("Method");

    // Rewrite bytecode; pad tail with nops.
    std::vector<uint8_t> padded(method.code_size, 0x00 /* nop */);
    std::memcpy(padded.data(), new_bytecode.data(), new_bytecode.size());
    r->write(method.code_base, padded.data(), padded.size());

    // Null out _code so any JIT-compiled body is dropped.
    uint64_t zero = 0;
    r->write(method.address + mt->field("_code")->offset, &zero, 8);

    // Retarget _from_compiled_entry / _from_interpreted_entry to the
    // interpreter-to-interpreter stub so calls re-enter the interpreter.
    if (mt->has_field("_i2i_entry")) {
        uint64_t i2i = r->read_u64(method.address + mt->field("_i2i_entry")->offset);
        if (i2i) {
            r->write(method.address + mt->field("_from_compiled_entry")->offset, &i2i, 8);
            r->write(method.address + mt->field("_from_interpreted_entry")->offset, &i2i, 8);
        }
    }
}

std::vector<uint8_t> make_nop_return()
{
    return std::vector<uint8_t>{0xB1}; // return (void)
}

} // namespace marrow

File: method_walker.py
"""Method walker + bytecode patcher.

Enumerates every Java method in an InstanceKlass and exposes just enough
of its structure to read its name/signature and rewrite its bytecode.

The bytecode-rewrite path underpins our invoke hack: we can't call
JavaCalls::call() from a non-JavaThread, but we can replace a method's
body and then let the target invoke that method naturally. JIT-compiled
versions are invalidated so the new bytecode actually runs.
"""
from __future__ import annotations

import struct
from dataclasses import dataclass

from vm_meta import VMMeta, _i32, _ptr, _u16
from field_reader import _ConstantPool


@dataclass(frozen=True)
class MethodSnapshot:
    address: int         # Method*
    const_method: int    # ConstMethod*
    name: str
    signature: str
    code_size: int       # length of the bytecode stream
    code_base: int       # absolute address of the first bytecode byte


def _array_methods(vm: VMMeta, klass_ptr: int) -> list[int]:
    """Return raw Method* pointers in InstanceKlass._methods."""
    r = vm.reader
    ik = vm.type("InstanceKlass")
    arr = _ptr(r, klass_ptr + ik.field("_methods").offset)
    if not arr:
        return []
    length = _i32(r, arr + 0)
    if length <= 0:
        return []
    raw = r.read(arr + 8, length * 8)
    return [struct.unpack_from("<Q", raw, i * 8)[0] for i in range(length)]


def methods_of(vm: VMMeta, klass_ptr: int) -> list[MethodSnapshot]:
    r = vm.reader
    method_t = vm.type("Method")
    cm_t = vm.type("ConstMethod")
    off_constmethod = method_t.field("_constMethod").offset
    off_name_idx = cm_t.field("_name_index").offset
    off_sig_idx = cm_t.field("_signature_index").offset
    off_code_size = cm_t.field("_code_size").offset
    off_cp = cm_t.field("_constants").offset
    header_size = cm_t.size  # bytecode starts right after the C++ header

    out: list[MethodSnapshot] = []
    for m_ptr in _array_methods(vm, klass_ptr):
        if not m_ptr:
            continue
        cm = _ptr(r, m_ptr + off_constmethod)
        if not cm:
            continue
        cp_ptr = _ptr(r, cm + off_cp)
        cp = _ConstantPool(vm, cp_ptr)
        name = cp.symbol(_u16(r, cm + off_name_idx))
        sig = cp.symbol(_u16(r, cm + off_sig_idx))
        code_size = _u16(r, cm + off_code_size)
        out.append(MethodSnapshot(
            address=m_ptr,
            const_method=cm,
            name=name,
            signature=sig,
            code_size=code_size,
            code_base=cm + header_size,
        ))
    return out


def find_method(vm: VMMeta, klass_ptr: int,
                name: str, signature: str | None = None
                ) -> MethodSnapshot | None:
    for m in methods_of(vm, klass_ptr):
        if m.name != name:
            continue
        if signature is not None and m.signature != signature:
            continue
        return m
    return None


# ---------------------------------------------------------------------------
# Bytecode patching
# ---------------------------------------------------------------------------

# Opcodes we care about:
#   0x00 — nop                    (single-byte padding)
#   0xB1 — return (void)          (returns from a V method)
#   0xAC — ireturn                (returns int from the top of stack)
#   0xAD — lreturn, 0xAE — freturn, 0xAF — dreturn, 0xB0 — areturn
_RETURN_VOID = 0xB1
_NOP = 0x00


def rewrite_method_body(
        vm: VMMeta, method: MethodSnapshot, new_bytecode: bytes) -> None:
    """Overwrite a method's bytecode in-place and invalidate the JIT version.

    `new_bytecode` must be <= method.code_size. The tail is padded with nops
    so interpreter dispatch never walks past the logical end.
    """
    if len(new_bytecode) > method.code_size:
        raise ValueError(
            f"new bytecode is {len(new_bytecode)}B but slot is "
            f"{method.code_size}B (can't grow without reallocating ConstMethod)")

    r = vm.reader
    method_t = vm.type("Method")

    # 1. Rewrite the bytecode stream. Padding matters because some code
    #    paths (e.g. verifier re-runs, reflection) walk the whole slot.
    padded = new_bytecode + bytes([_NOP] * (method.code_size - len(new_bytecode)))
    r.write(method.code_base, padded)

    # 2. Invalidate any JIT-compiled body: null out `_code` and reset the
    #    entry trampolines to the interpreter-to-interpreter stub.
    #    Without this, the target would keep calling the old compiled form.
    r.write(method.address + method_t.field("_code").offset,
            struct.pack("<Q", 0))

    if method_t.has_field("_i2i_entry"):
        i2i = _ptr(r, method.address + method_t.field("_i2i_entry").offset)
        if i2i:
            fci_off = method_t.field("_from_compiled_entry").offset
            fie_off = method_t.field("_from_interpreted_entry").offset
            r.write(method.address + fci_off, struct.pack("<Q", i2i))
            r.write(method.address + fie_off, struct.pack("<Q", i2i))


def make_nop_return(method: MethodSnapshot) -> bytes:
    """Build bytecode that immediately returns (for a V-type method).

    Caller must confirm the method signature matches the return opcode;
    only void methods may use `return` (0xB1)."""
    return bytes([_RETURN_VOID])
